grandpa/primitives/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2023 Polytope Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Primitive types and traits used by the GRANDPA prover & verifier.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::all)]
+#![deny(missing_docs)]
+
+extern crate alloc;
+
+use alloc::collections::BTreeMap;
+use codec::{Decode, Encode};
+use core::fmt::Debug;
+use sp_core::{ed25519, sp_std, H256};
+use sp_finality_grandpa::{AuthorityId, AuthorityList, AuthoritySignature};
+use sp_runtime::traits::Header;
+use sp_std::prelude::*;
+use sp_storage::StorageKey;
+
+/// GRANPA errors
+pub mod error;
+/// GRANDPA justification utilities
+pub mod justification;
+pub mod host_functions;
+
+/// Represents a Hash in this library
+pub type Hash = H256;
+/// A commit message for this chain's block type.
+pub type Commit<H> = finality_grandpa::Commit<
+    <H as Header>::Hash,
+    <H as Header>::Number,
+    AuthoritySignature,
+    AuthorityId,
+>;
+
+/// Finality for block B is proved by providing:
+/// 1) the justification for the descendant block F;
+/// 2) headers sub-chain (B; F] if B != F;
+#[derive(Debug, PartialEq, Encode, Decode, Clone)]
+pub struct FinalityProof<H: codec::Codec> {
+    /// The hash of block F for which justification is provided.
+    pub block: Hash,
+    /// Justification of the block F.
+    pub justification: Vec<u8>,
+    /// The set of headers in the range (B; F] that we believe are unknown to the caller. Ordered.
+    pub unknown_headers: Vec<H>,
+}
+
+/// Previous light client state.
+#[derive(Clone)]
+pub struct ConsensusState {
+    /// Current authority set
+    pub current_authorities: AuthorityList,
+    /// Id of the current authority set.
+    pub current_set_id: u64,
+    /// latest finalized height on the relay chain.
+    pub latest_relay_height: u32,
+    /// latest finalized height on the parachain.
+    pub latest_para_height: u32,
+    /// latest finalized hash on the relay chain.
+    pub latest_relay_hash: Hash,
+    /// para_id of associated parachain
+    pub para_id: u32,
```

Comment:
I think this should be a map, `BtreeMap<u32, bool>`. So that the verifier only accepts parachain headers listed in this map